File: backend/src/main/scala/slaydemo/backend/battle/objects/BattlePickupAvailability.scala
package slaydemo.backend.battle.objects

enum BattlePickupAvailability {
  case Available
  case Respawning(remainingMs: DurationMillis)
}

object BattlePickupAvailability {
  def availableFlag(value: BattlePickupAvailability): Boolean =
    value == BattlePickupAvailability.Available

  def respawnMs(value: BattlePickupAvailability): DurationMillis =
    value match {
      case BattlePickupAvailability.Available =>
        DurationMillis(0L)
      case BattlePickupAvailability.Respawning(remainingMs) =>
        DurationMillis(math.max(0L, remainingMs.value))
    }

  def respawning(remainingMs: DurationMillis): BattlePickupAvailability =
    if remainingMs.value <= 0L then BattlePickupAvailability.Available
    else BattlePickupAvailability.Respawning(DurationMillis(remainingMs.value))

  def fromAvailableFlag(available: Boolean, respawnMs: DurationMillis): BattlePickupAvailability =
    if available then BattlePickupAvailability.Available
    else BattlePickupAvailability.Respawning(DurationMillis(math.max(0L, respawnMs.value)))
}

File: frontend/src/objects/battle/microservices/abilities/objects/pickup/BattlePickupAvailability.ts
export type BattlePickupAvailability =
  | { status: "available" }
  | { status: "respawning"; remainingMs: number };

export function availableFlag(value: BattlePickupAvailability): boolean {
  return value.status === "available";
}

export function respawnMs(value: BattlePickupAvailability): number {
  return value.status === "respawning" ? Math.max(0, Math.round(value.remainingMs)) : 0;
}

export function respawning(remainingMs: number): BattlePickupAvailability {
  return remainingMs <= 0 ? { status: "available" } : { status: "respawning", remainingMs: Math.round(remainingMs) };
}

export function fromAvailableFlag(available: boolean, remainingMs: number): BattlePickupAvailability {
  return available ? { status: "available" } : respawning(remainingMs);
}

